core/src/main/java/org/apache/calcite/util/format/postgresql/PostgresqlDateTimeFormatter.java

```diff
@@ -308,4 +481,362 @@ public static String toChar(String formatString, ZonedDateTime dateTime) {
+    }
+
+    return updatedDateTime;
+  }
+
+  private static int getGregorianYear(Map<ChronoUnitEnum, Long> dateParts) {
+    return getYear(
+        dateParts.get(ChronoUnitEnum.ERAS),
+        dateParts.get(ChronoUnitEnum.YEARS),
+        dateParts.get(ChronoUnitEnum.CENTURIES),
+        dateParts.get(ChronoUnitEnum.YEARS_IN_MILLENIA),
+        dateParts.get(ChronoUnitEnum.YEARS_IN_CENTURY));
+  }
+
+  private static LocalDateTime updateWithIso8601Fields(LocalDateTime dateTime,
+      Map<ChronoUnitEnum, Long> dateParts) {
+    final int year = getIso8601Year(dateParts);
+
+    if (!dateParts.containsKey(ChronoUnitEnum.WEEKS_IN_YEAR_ISO_8601)
+        && !dateParts.containsKey(ChronoUnitEnum.DAYS_IN_YEAR_ISO_8601)) {
+      return dateTime.withYear(year).withDayOfYear(1);
+    }
+
+    LocalDateTime updatedDateTime = dateTime
+        .with(ChronoField.DAY_OF_WEEK, 1)
+        .with(IsoFields.WEEK_BASED_YEAR, year)
+        .with(IsoFields.WEEK_OF_WEEK_BASED_YEAR, 1);
+
+    if (dateParts.containsKey(ChronoUnitEnum.WEEKS_IN_YEAR_ISO_8601)) {
+      updatedDateTime =
+          updatedDateTime.with(IsoFields.WEEK_OF_WEEK_BASED_YEAR,
+              dateParts.get(ChronoUnitEnum.WEEKS_IN_YEAR_ISO_8601));
+
+      if (dateParts.containsKey(ChronoUnitEnum.DAYS_IN_WEEK)) {
+        updatedDateTime =
+            updatedDateTime.with(ChronoField.DAY_OF_WEEK,
+                dateParts.get(ChronoUnitEnum.DAYS_IN_WEEK));
+      }
+    } else if (dateParts.containsKey(ChronoUnitEnum.DAYS_IN_YEAR_ISO_8601)) {
+      updatedDateTime =
+          updatedDateTime.plusDays(dateParts.get(ChronoUnitEnum.DAYS_IN_YEAR_ISO_8601) - 1);
+    }
+
+    return updatedDateTime;
+  }
+
+  private static int getIso8601Year(Map<ChronoUnitEnum, Long> dateParts) {
+    return getYear(
+        dateParts.get(ChronoUnitEnum.ERAS),
+        dateParts.get(ChronoUnitEnum.YEARS_ISO_8601),
+        dateParts.get(ChronoUnitEnum.CENTURIES),
+        dateParts.get(ChronoUnitEnum.YEARS_IN_MILLENIA_ISO_8601),
+        dateParts.get(ChronoUnitEnum.YEARS_IN_CENTURY_ISO_8601));
+  }
+
+  private static int getYear(@Nullable Long era, @Nullable Long years,
+      @Nullable Long centuries, @Nullable Long yearsInMillenia,
+      @Nullable Long yearsInCentury) {
+    int yearSign = 1;
+    if (era != null) {
+      if (era == 0) {
+        yearSign = -1;
+      }
+    }
+
+    if (yearsInMillenia != null) {
+      int year = yearsInMillenia.intValue();
+      if (year < 520) {
+        year += 2000;
+      } else {
+        year += 1000;
+      }
+
+      return yearSign * year;
+    }
+
+    if (centuries != null) {
+      int year = 100 * (centuries.intValue() - 1);
+
+      if (yearsInCentury != null) {
+        year += yearsInCentury.intValue();
+      } else {
+        year += 1;
+      }
+
+      return yearSign * year;
+    }
+
+    if (years != null) {
+      return yearSign * years.intValue();
+    }
+
+    if (yearsInCentury != null) {
+      int year = yearsInCentury.intValue();
```

Comment:
is this part of the postgres spec?

Reply:
Yeah, `YY`. You could do something like `CCYY` if you wanted to enter this year as `2124`.

```
postgres=# select to_date('2124', 'CCYY');
  to_date
------------
 2024-01-01
(1 row)
```